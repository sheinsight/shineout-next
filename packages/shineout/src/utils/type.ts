export type RequireAWithB<T, A extends keyof T, B extends keyof T> = 
  | (Omit<T, A> & { [K in A]?: never })
  | (T & Required<Pick<T, B>>)
  | (T & Required<Pick<T, A | B>>)