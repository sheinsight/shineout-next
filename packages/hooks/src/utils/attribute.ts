export const getDataAttribute = (attrs: Record<string, string>) => {
  return Object.keys(attrs).reduce((acc, key) => {
    const ns = `data-soui-${key}`;
    acc[ns] = attrs[key];
    return acc;
  }, {} as Record<string, string>);
};
