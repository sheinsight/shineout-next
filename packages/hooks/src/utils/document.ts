import { isBrowser } from './is';
export const docSize = {
  get width() {
    if (!isBrowser()) return 0;
    return document.documentElement.clientWidth || document.body.clientWidth;
  },
  get height() {
    if (!isBrowser()) return 0;
    return document.documentElement.clientHeight || document.body.clientHeight;
  },
};
export const docScroll = {
  get top() {
    if (!isBrowser()) return 0;
    return document.documentElement.scrollTop || document.body.scrollTop;
  },
  get left() {
    if (!isBrowser()) return 0;
    return document.documentElement.scrollLeft || document.body.scrollLeft;
  },
  set top(value) {
    if (!isBrowser()) return;
    document.documentElement.scrollTop = value;
    document.body.scrollTop = value;
  },
  set left(value) {
    if (!isBrowser()) return;
    document.documentElement.scrollLeft = value;
    document.body.scrollLeft = value;
  },
};

export function isInDocument(element: HTMLElement | null) {
  if (element && 'isConnected' in element) {
    return element.isConnected;
  }
  return isBrowser() && document.documentElement.contains(element);
}
