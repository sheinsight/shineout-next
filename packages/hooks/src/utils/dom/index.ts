export { default, default as NormalizeWheel } from './normalize-wheel';
export type { OldWheelEvent } from './normalize-wheel.type';
