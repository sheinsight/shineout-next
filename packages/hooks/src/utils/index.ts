export * from './is';
export { default as extractEventHandlers } from './extractEventHandlers';
export * from './clone';
export * from './object';
export * from './string';
export * from './func';
export * from './validate';
export * from './number';
export * from './shallowEqual';
export * from './uid';
export * from './document';
export * from './position';
export * from './render';
export * from './lazyload';
export * from './react-render';
export * from './style-unit';
export * from './attribute';
