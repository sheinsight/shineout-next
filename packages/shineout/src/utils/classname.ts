import classnames from 'classnames';
import { config } from '@sheinx/base';

interface CSSModule {
  [key: string]: string;
}

/**
 * create a new className generate function, add namespace, handle css module
 * @param style - object; for css module
 * @param module - string
 * @param prefix - string, default value is 'shineout'
 * * */
export default (style: CSSModule, module: string, prefix: string = config.prefix) =>
  (...args: any) => {
    const className = classnames(...args);
    if (!className) return '';

    const ns = `${prefix}${module ? `-${module}` : '-'}`;
    let list: (string | number)[] = className
      .split(' ')
      .map((c) => (c === '_' ? ns : `${ns}-${c}`));
    if ((config as any).cssModule) {
      list = list.map((c: any) => style[c] || c);
    }
    return list.join(' ');
  };
