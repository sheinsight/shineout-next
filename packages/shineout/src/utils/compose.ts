export function compose(...funcs: any) {
  if (funcs.length === 0) {
    return (arg: any) => arg;
  }
  const last = funcs[funcs.length - 1];
  const rest = funcs.slice(0, -1);
  return (...args: typeof last) =>
    rest.reduceRight((composed: any, f: any) => f(composed), last(...args));
}
