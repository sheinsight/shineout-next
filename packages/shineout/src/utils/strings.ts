import { util } from "@sheinx/hooks";

const { devUseWarning } = util;

export function capitalize(str: string) {
  if (typeof str !== 'string') {
    devUseWarning.error('str should be a string');
  }
  return str && str[0].toUpperCase() + str.slice(1);
}

export function substitute(str: string | Function, obj: { [x: string]: any }) {
  if (typeof str === 'string') {
    if (str.indexOf('{') < 0) {
      return str;
    }

    return str.replace(/\\?\{([^{}]+)\}/g, (match, name) => {
      if (match.charAt(0) === '\\') {
        return match.slice(1);
      }
      return obj[name] === null || obj[name] === undefined ? '' : obj[name];
    });
  }
  if (typeof str === 'function') {
    let val = str(obj);
    if (val === obj && typeof val === 'object') {
      val = Object.assign({}, obj);
    }
    return val;
  }

  return '';
}

export function removeProtocol(url: string) {
  if (url.indexOf('http') !== 0) return url;
  try {
    const { href, protocol } = new URL(url);
    return href.slice(protocol.length);
  } catch (error) {
    return url;
  }
}

export function getRTLPosition(position: string) {
  if (!position) return position;
  // position.replace('left', 'right').replace('right', 'left')
  if (position.indexOf('left') !== -1) {
    return position.replace('left', 'right');
  }
  if (position.indexOf('right') !== -1) {
    return position.replace('right', 'left');
  }
  return position;
}

export function getDirectionIconName(mode = 'left', double = false) {
  const rtl = true;
  if (mode === 'left') {
    if (rtl) {
      return double ? 'ArrowRightDouble' : 'AngleRight';
    }
    return double ? 'ArrowLeftDouble' : 'AngleLeft';
  }

  if (rtl) {
    return double ? 'ArrowLeftDouble' : 'AngleLeft';
  }
  return double ? 'ArrowRightDouble' : 'AngleRight';
}
