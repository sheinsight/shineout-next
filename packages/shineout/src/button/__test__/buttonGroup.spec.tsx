import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import Button from '..';
import { classTest, createClassName } from '../../tests/utils';
import ButtonGroup from '../__example__/s-008-group';

const SO_PREFIX = 'button';
const originClasses = ['group'];
const originItemClasses = ['secondary', 'outline', 'primary', 'text', 'small', 'round'];
const { group, secondary, outline, text, small, round } = createClassName(
  SO_PREFIX,
  originClasses,
  originItemClasses,
);

afterEach(cleanup);
describe('ButtonGroup[Base]', () => {
  test('should render correctly', () => {
    const { container } = render(<ButtonGroup />);
    expect(container.firstChild).toMatchSnapshot();
  });
  test('should render button Group base', () => {
    const { container } = render(<ButtonGroup />);
    expect(container.querySelectorAll(group).length).toBe(5);
  });
  test('should render when set type and mode in button group', () => {
    const { container } = render(
      <Button.Group type='secondary' mode='outline'>
        <Button>test</Button>
        <Button>test</Button>
      </Button.Group>,
    );
    expect(container.querySelectorAll(group).length).toBe(1);
    container.querySelectorAll('button').forEach((button) => {
      expect(button.classList.contains(secondary)).toBeTruthy();
      expect(button.classList.contains(outline)).toBeTruthy();
    });
  });
  test('should render children button type when set type in group and button at the same time', () => {
    const { container } = render(
      <Button.Group type='primary'>
        <Button type='secondary'>secondary</Button>
        <Button type='success'>success</Button>
      </Button.Group>,
    );
    screen.debug();
    container.querySelectorAll('button').forEach((button) => {
      expect(button.classList[0].includes(button.textContent as string)).toBeTruthy();
    });
  });
  test('should render group mode when set mode in group and button at the same time', () => {
    const { container } = render(
      <Button.Group mode='text'>
        <Button mode='link'>link</Button>
        <Button mode='outline'>outline</Button>
      </Button.Group>,
    );
    container.querySelectorAll('button').forEach((button) => {
      classTest(button, text);
    });
  });
  test('should render when set size in group and button at the same time', () => {
    const { container } = render(
      <Button.Group size='small'>
        <Button size='default'>default</Button>
        <Button size='large'>large</Button>
      </Button.Group>,
    );
    expect(container.querySelector(group)?.classList.contains(small)).toBeTruthy();
    container.querySelectorAll('button').forEach((button) => {
      expect(button.classList.contains(small)).toBeTruthy();
    });
  });
  test('should render when set shape in group and button at the same time', () => {
    const { container } = render(
      <Button.Group shape='round'>
        <Button shape='circle'>circle</Button>
        <Button shape='square'>square</Button>
        <Button>test</Button>
      </Button.Group>,
    );
    expect(container.querySelector(group)?.classList.contains(round)).toBeTruthy();
    container.querySelectorAll('button').forEach((button) => {
      expect(button.classList.contains(round)).toBeTruthy();
    });
  });
  // TODO: Group上的属性会透传
  test('should render when children is not button', () => {
    const { container } = render(
      <Button.Group shape='round'>
        <div>test</div>
      </Button.Group>,
    );
    screen.debug();
    expect(container.querySelector(group)?.classList.contains(round)).toBeTruthy();
  });
  test('should not click when set click in button group', () => {
    const handleFn = jest.fn();
    const { container } = render(
      <Button.Group onClick={handleFn}>
        <Button>test</Button>
      </Button.Group>,
    );
    fireEvent.click(container.querySelector(group) as Element);
    expect(handleFn.mock.calls.length).toBe(0);
    fireEvent.click(container.querySelector('button') as Element);
    expect(handleFn.mock.calls.length).toBe(0);
  });
  test('should click when set click in button', () => {
    const handleFnA = jest.fn();
    const handleFnB = jest.fn();
    const { container } = render(
      <Button.Group>
        <Button onClick={handleFnA}>test</Button>
        <Button onClick={handleFnB}>test</Button>
      </Button.Group>,
    );
    fireEvent.click(container.querySelector(group) as Element);
    expect(handleFnA.mock.calls.length).toBe(0);
    expect(handleFnB.mock.calls.length).toBe(0);
    fireEvent.click(container.querySelectorAll('button')[0] as Element);
    expect(handleFnA.mock.calls.length).toBe(1);
    fireEvent.click(container.querySelectorAll('button')[1] as Element);
    expect(handleFnB.mock.calls.length).toBe(1);
  });
});
