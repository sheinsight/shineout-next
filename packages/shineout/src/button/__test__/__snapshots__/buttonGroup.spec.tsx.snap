// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup[Base] should render correctly 1`] = `
<div
  style="display: flex; gap: 24px; flex-wrap: wrap; flex-direction: column;"
>
  <div
    class="so-button-group"
  >
    <button
      class="so-button-primary so-button-button so-button-text"
    >
      <span>
        Publish
      </span>
    </button>
    <button
      class="so-button-secondary so-button-button so-button-text"
    >
      <span>
        Publish
      </span>
    </button>
    <button
      class="so-button-secondary so-button-button so-button-text"
    >
      <span>
        Publish
      </span>
    </button>
    <button
      class="so-button-secondary so-button-button so-button-disabled so-button-text"
      disabled=""
    >
      <span>
        Publish
      </span>
    </button>
  </div>
  <div
    class="so-button-group"
  >
    <button
      class="so-button-primary so-button-button"
    >
      <span>
        Publish
      </span>
    </button>
    <button
      class="so-button-secondary so-button-button"
    >
      <span>
        Publish
      </span>
    </button>
    <button
      class="so-button-secondary so-button-button"
    >
      <span>
        Publish
      </span>
    </button>
    <button
      class="so-button-secondary so-button-button"
    >
      <span>
        Publish
      </span>
    </button>
  </div>
  <div
    class="so-button-group"
  >
    <button
      class="so-button-secondary so-button-button so-button-outline"
    >
      <span>
        Publish
      </span>
    </button>
    <button
      class="so-button-secondary so-button-button so-button-outline"
    >
      <span>
        Publish
      </span>
    </button>
    <button
      class="so-button-secondary so-button-button so-button-outline"
    >
      <span>
        Publish
      </span>
    </button>
    <button
      class="so-button-secondary so-button-button so-button-outline"
    >
      <span>
        Publish
      </span>
    </button>
  </div>
  <div
    style="display: flex; gap: 24px;"
  >
    <div
      class="so-button-group"
    >
      <button
        class="so-button-primary so-button-button"
      >
        <span>
          Publish
        </span>
      </button>
      <button
        class="so-button-primary so-button-button"
      >
        <span
          style="display: inline-flex; line-height: 22px;"
        >
          <svg
            fill="none"
            height="14"
            viewBox="0 0 24 24"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.2929 15.7071C11.6534 16.0676 12.2206 16.0953 12.6129 15.7903L12.7071 15.7071L18.7071 9.70711C19.0976 9.31658 19.0976 8.68342 18.7071 8.29289C18.3466 7.93241 17.7794 7.90468 17.3871 8.2097L17.2929 8.29289L12.7072 12.8779C12.3165 13.2683 11.6835 13.2683 11.2928 12.8779L6.70711 8.29289C6.34662 7.93241 5.77939 7.90468 5.3871 8.2097L5.29289 8.29289C4.93241 8.65338 4.90468 9.22061 5.2097 9.6129L5.29289 9.70711L11.2929 15.7071Z"
              fill="#ffffff"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="so-button-group"
    >
      <button
        class="so-button-secondary so-button-button"
      >
        <span>
          Publish
        </span>
      </button>
      <button
        class="so-button-secondary so-button-button"
      >
        <span
          style="display: inline-flex; line-height: 22px;"
        >
          <svg
            fill="none"
            height="14"
            viewBox="0 0 24 24"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 10C6.10457 10 7 10.8954 7 12C7 13.1046 6.10457 14 5 14C3.89543 14 3 13.1046 3 12C3 10.8954 3.89543 10 5 10ZM12 10C13.1046 10 14 10.8954 14 12C14 13.1046 13.1046 14 12 14C10.8954 14 10 13.1046 10 12C10 10.8954 10.8954 10 12 10ZM19 10C20.1046 10 21 10.8954 21 12C21 13.1046 20.1046 14 19 14C17.8954 14 17 13.1046 17 12C17 10.8954 17.8954 10 19 10Z"
              fill="#666C7C"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
